File: bin/client/dotatribe/resource/script/view/IWindow.lua
--Name : IView.lua
--Function : 提供Lua页面基类和页面管理
--Author : fangjun

require "resource/script/util/util_function"
require "resource/script/event/EventTyper"
require "resource/script/event/EventHelper"
require "resource/script/view/UIWindow"
require "resource/script/view/UIHelper"

IWindow = {}

IWindow.m_WindowType = 0
IWindow.m_WindowName = ""
IWindow.m_bModelWindow = false
IWindow.m_iWindowZOrder = 0
IWindow.m_bLoaded = false
IWindow.m_bShow = false
IWindow.m_pRootNode = nil
IWindow.m_bIsForceClearUnUseTex = false
IWindow.m_bIsForceClearAnimation = false
IWindow.m_iSoundOpenId = 0 -- 此页面打开音效
IWindow.m_iSoundCloseId = 0 -- 此页面关闭音效

-- 强推送事件列表(用于推送事件时)
IWindow.m_hardEventTable = {}

-- 窗口构造方法
function IWindow:New(windowType)
	local object = object or {}
	setmetatable(object, self)
	self.__index = self
	self.super = self

	object.m_WindowType = windowType
	return object
end


function IWindow:GetWindowType()
	return self.m_WindowType
end

function IWindow:GetWindowName()
	return self.m_WindowName
end

function IWindow:IsLoad()
	return self.m_bLoaded
end

function IWindow:GetRootNode()
	return self.m_pRootNode
end

function IWindow:Show(bshow)
	print("IWindow:Show   begin")
	if self.m_bShow ~= bshow then
		self.m_bShow = bshow
		if self.m_pRootNode ~= nil then
			if bshow == true then
				self.m_pRootNode:setVisible(true)
				-- 播放打开声音
				self:PlayOpenSound()
			else
				self.m_pRootNode:setVisible(false)
				--播放关闭声音
				self:PlayCloseSound()
			end
		end
	end
	print("IWindow:Show   end")
end

function IWindow:IsVisible()
	return self.m_bShow
end

function IWindow:GetPosition()
	if self.m_pRootNode ~= nil then
		return self.m_pRootNode:GetPosition()
	else
		return 0, 0
	end
end

function IWindow:SetPosition(x, y)
	if self.m_pRootNode ~= nil then
		self.m_pRootNode:SetPosition(x, y)
	end
end

function IWindow:RegisterEvent(eventId, bNotifyOnlyVisible)
	if bNotifyOnlyVisible == nil then
		table.insert(self.m_hardEventTable, eventId)

	end

	-- 调用EventHelper或者其他模块将事件注册进喜用eventsystem
	log("aaaa windowtype " .. eventId .. " windowtype " .. self.m_WindowType)
	Call_EventHelper():RegisterEvent(eventId, self.m_WindowType)
end

-- 消息处理，，此函数需要子类实现
function IWindow:OnEventHandler(eventId, paramTable)
end

function IWindow:PlayOpenSound()
	local id = "" .. self.m_iSoundOpenId

	-- 调用EventHelper Push机制将事件和参数推出
	Call_EventHelper():PushEvent(381, id)
end

function IWindow:PlayCloseSound()
	local id = "" .. self.m_iSoundCloseId

	-- 调用EventHelper Push机制将事件和参数推出
	Call_EventHelper():PushEvent(381, id)
end

function IWindow:IsHardEvent(eventId)
	for item in list_iter(self.m_hardEventTable) do
		log("Is hard Event  " .. eventId .. "..." .. item)

		if item == eventId then
			return true
		end
	end
	return false
end

function IWindow:Init()
	-- 需要从UIWindow.txt中获取数据进行配置
	log("IWindow:Init  begin")

	local uiwindowdata = Call_UIHelper():GetUIWindowData(self:GetWindowType())
	self.m_WindowName = uiwindowdata.name
	self.m_bModelWindow = uiwindowdata.isModel
	self.m_iWindowZOrder = uiwindowdata.ZOrder
	self.m_bLoaded = false
	self.m_bShow = false
	self.m_pRootNode = nil
	self.m_bIsForceClearUnUseTex = true
	self.m_bIsForceClearAnimation = true
	self.m_iSoundOpenId = uiwindowdata.openSound -- 此页面打开音效
	self.m_iSoundCloseId = uiwindowdata.closeSound -- 此页面关闭音效

	log("IWindow:Init  end")
end

function IWindow:Load()
	log("IWindow:Load   begin")
	self.m_pRootNode = CCNode:create()
	self.m_pRootNode:setPosition(ccp(0, 0))

	--self.m_pRootNode:setZOrder(self.m_iWindowZOrder)
	local layer = Call_UIHelper():GetUILayer()
	layer:addChild(self.m_pRootNode, self.m_iWindowZOrder)

	if self.m_bModelWindow then
		local director = CCDirector:sharedDirector()
		--director:getTouchDispatcher:addTargetedDelegate()
	end

	self.m_bLoaded = true

	log("IWindow:Load   end")

	log("Load UI Window, WindowType=" .. self:GetWindowType() .. ", WindowName=" .. self:GetWindowName())
	return true
end

function IWindow:Tick()

end

function IWindow:Destroy()
	log("IWindow:Destroy begin")

	self:PlayCloseSound()

	if self.m_pRootNode ~= nil then
		self.m_pRootNode:removeFromParentAndCleanup(true)
		--self.m_pRootNode:removeAllChildrenWithCleanup(true)
		--Call_UIHelper():removeChild(self.m_pRootNode, true)

		self.m_pRootNode = nil
	end

	self.m_bShow = false
	self.m_bLoaded = false

	if self.m_bIsForceClearAnimation == true then
		-- 强制清除动画,调用
		Call_UIHelper():ForceClearAnimationCache()
	end



	log("IWindow:Destroy end")

	log("UnLoad UI Window, WindowType=" .. self:GetWindowType() .. ", WindowName=" .. self:GetWindowName())
end

File: src/project/DOTATribe/src/IWindow.cpp
#include "../include/IWindow.h"
#include "../include/UISystem.h"
#include "../include/LoggerSystem.h"
#include "../include/EngineSystem.h"
#include "../include/SoundSystem.h"
#include "../include/UIWindowDataManager.h"
#include "../include/WindowManager.h"
#include "../include/EventSystem.h"
#include "../include/EventTyper.h"
#include "../include/DOTATribeApp.h"
#include "../include/TimeSystem.h"
#include "../include/config_SoundSystem.h"
#include <cocos2d.h>
#include <string>

IWindow::IWindow()
{
	m_WindowType   = _TYPED_WINDOW_UNKOWN_;
	m_WindowName   = "";
	m_bModelWindow = false;
	m_WindowZOrder = _TYPED_WINDOW_ZORDER_NORMAL_;
	m_bLoaded      = false;
	m_bShow        = false;
	m_pRootNode    = NULL;
	m_bTickListener= true;
	m_TickCounter  = 0;
	m_bIsForceClearUnUseTex=true;
	m_bIsForceClearAnimation=true;
	m_ParentWindowName = "root";
}

IWindow::~IWindow()
{

}

int IWindow::GetWindowType()
{
	return m_WindowType;
}

bool IWindow::IsLoad()
{	
	return m_bLoaded;
}

cocos2d::CCNode* IWindow::GetRootNode()
{
	return m_pRootNode;
}

void IWindow::SetWindowDataByPos(std::string str, unsigned int pos)
{
	m_WindowDataList[pos] = str;
}

std::string IWindow::GetWindowDataByPos(unsigned int pos)
{
	return m_WindowDataList[pos];
}

void IWindow::SetTickListenser(bool bListenser)
{
	m_bTickListener = bListenser;
}

void IWindow::Show(bool bShow)
{

	if (m_bShow == bShow)
		return ;

	m_bShow = bShow;
	if (m_pRootNode == NULL)
		return ;

	if (bShow)
	{
		m_pRootNode->setVisible(true);
		PlayOpenSound();
	}
	else
	{
		m_pRootNode->setVisible(false);
		PlayCloseSound();
	}
}

bool IWindow::IsVisible()
{
	return m_bShow;
}

void IWindow::SetXPosition(float xPos)
{
	if (m_pRootNode == NULL)
		return ;

	m_pRootNode->setPositionX(xPos);
}

float IWindow::GetXPosition()
{
	if (m_pRootNode == NULL)
		return 0.0f;

	return m_pRootNode->getPositionX();
}

void IWindow::SetYPosition(float yPos)
{
	if (m_pRootNode == NULL)
		return ;

	m_pRootNode->setPositionY(yPos);
}

float IWindow::GetYPosition()
{
	if (m_pRootNode == NULL)
		return 0.0f;

	return m_pRootNode->getPositionY();
}

void IWindow::SetXScale(float xScale)
{
	if (m_pRootNode == NULL)
		return ;

	m_pRootNode->setScaleX(xScale);
}

void IWindow::SetYScale(float yScale)
{
	if (m_pRootNode == NULL)
		return ;

	m_pRootNode->setScaleY(yScale);
}

void IWindow::RegisterEvent(int iEventType, bool bNotifyOnlyWhenVisible)
{
	if (!bNotifyOnlyWhenVisible)
		UISYSTEM->RegisterUIEvent(iEventType, m_WindowType);
	else
		UISYSTEM->RegisterUIEvent2(iEventType, m_WindowType);
}

void IWindow::PushEvent(int iEventType)
{
	UISYSTEM->PushEvent(iEventType);
}

void IWindow::PushEvent(int iEventType, char* pArgs)
{
	assert(pArgs != NULL);

	UISYSTEM->PushEvent(iEventType, std::string(pArgs));
}

void IWindow::PlayOpenSound()
{
	UIWindowItemData* pItem = UIWindowDataManager::get_instance2()->GetWindowByType(m_WindowType);
	if (pItem==NULL || pItem->mSoundOpen == -1)
		return ;

	if (m_WindowType != _TYPED_WINDOW_ATHLETICSRESULT_)
	{
		char _soundID[64]="";
		sprintf(_soundID,"%d",pItem->mSoundOpen);
		EVENTSYSTEM->PushEvent(_TYPED_EVENT_SOUND_PLAY_,_soundID);
	}
}

void IWindow::PlayCloseSound()
{
	UIWindowItemData* pItem = UIWindowDataManager::get_instance2()->GetWindowByType(m_WindowType);
	if (pItem==NULL || pItem->mSoundClose == -1)
		return ;

	if (m_WindowType != _TYPED_WINDOW_ATHLETICSRESULT_)
	{
		char _soundID[64]="";
		sprintf(_soundID,"%d",pItem->mSoundClose);
		EVENTSYSTEM->PushEvent(_TYPED_EVENT_SOUND_PLAY_,_soundID);
	}
}
void IWindow::RegisterTimerFunction(float time, std::string functionName)
{
	_SCRIPT_TIMER_FUNCTION_ func;
	func.mStartTime     = m_TickCounter;
	func.mIntervalTime  = int(time*DOTATribeApp::get_instance2()->GetFPS());
	func.mScriptFunction= functionName;
}

bool IWindow::Init()
{
	UIWindowItemData* pWindowItem = UIWindowDataManager::get_instance2()->GetWindowByType(this->GetWindowType());
	m_bModelWindow = pWindowItem->mIsModelWindow;
	m_WindowZOrder = pWindowItem->mWindowZOrder;
	m_WindowName   = pWindowItem->mWindowName;

	LOGGERSYSTEM->LogInfo("Init UI Window, WindowType=%d, WindowName=%s\n", m_WindowType, m_WindowName.c_str());
	return true;
}

bool IWindow::Load()
{
	do 
	{
		//这一帧取消所有按钮事件==CPriorityMenu
		//下一帧恢复
		UISYSTEM->Set_Global_menuitem_enabled(false);
		//创建根节点
		m_pRootNode=CCNode::create();
		m_pRootNode->setPosition(CCPointZero);
		UISYSTEM->GetUILayer()->addChild(m_pRootNode, m_WindowZOrder);
		m_pRootNode->setVisible(false);
		// 注册消息处理事件
		if (m_bModelWindow)
		{
			cocos2d::CCDirector::sharedDirector()->getTouchDispatcher()->addTargetedDelegate(this, m_pRootNode->getPriority(), true);
		}

		//加载资源Plist
		LoadPlistForWindow();

		m_bLoaded = true;
	} 
	while (false);

	LOGGERSYSTEM->LogInfo("Load UI Window, WindowType=%d, WindowName=%s\n", m_WindowType, m_WindowName.c_str());
	return true;
}

void IWindow::Tick()
{
	m_TickCounter++;
}

void IWindow::Destroy()
{
	do 
	{
		PlayCloseSound();
		// 删除所有该窗口绑定的c++定时回调函数
		m_bRunning = false;
		unscheduleAllSelectors();
		// 卸载消息处理事件
		if (m_bModelWindow)
		{
			cocos2d::CCDirector::sharedDirector()->getTouchDispatcher()->removeDelegate(this);
		}
		// 从父节点删除当前窗口
		if (m_pRootNode != NULL)
		{
			m_pRootNode->removeAllChildrenWithCleanup(true);
			UISYSTEM->GetUILayer()->removeChild(m_pRootNode, true);
			m_pRootNode = NULL;
		}

		m_bShow   = false;
		m_bLoaded = false;
	}
	while (false);

	//清除资源plist
	UnloadPlistForWindow();

	if(m_bIsForceClearAnimation)
	{
		DOTATribeApp::get_instance2()->ForceClearAnimationCache();
	}
	
	UISystem::get_instance2()->Set_Global_UIClear_Enable(true);

	LOGGERSYSTEM->LogInfo("UnLoad UI Window, WindowType=%d, WindowName=%s\n", m_WindowType, m_WindowName.c_str());
}

	/*
		加载窗口所需图片资源plist
	*/
void IWindow::LoadPlistForWindow(){
	//加载指定资源Plist
	UIWindowItemData* pWindowItem = UIWindowDataManager::get_instance2()->GetWindowByType(this->GetWindowType());
	if (strlen(pWindowItem->mBGImageResourceName)>0)
	{
		_TRACE_LOADING_PLIST_(pWindowItem->mBGImageResourceName)
	}
	if (strlen(pWindowItem->mUIResourceName)>0)
	{
		_TRACE_LOADING_PLIST_(pWindowItem->mUIResourceName)	
	}
}
	/*
		清除窗口所需图片资源plist
	*/
void IWindow::UnloadPlistForWindow(){
	//强行清除图片资源
	if(m_bIsForceClearAnimation)
		DOTATribeApp::get_instance2()->ForceClearAnimationCache();
	if(m_bIsForceClearUnUseTex)
		DOTATribeApp::get_instance2()->ForceFreeMemory(); 
	//卸载指定资源Plist
	UIWindowItemData* pWindowItem = UIWindowDataManager::get_instance2()->GetWindowByType(this->GetWindowType());
	if (pWindowItem->mIsBGImageReleaseWhenWindowDestroy&&strlen(pWindowItem->mBGImageResourceName)>0)
	{
		_TRACE_UNLOAD_PLIST_(pWindowItem->mBGImageResourceName)
	}
	if (pWindowItem->mIsUIImageReleaseWhenWindowDestroy&&strlen(pWindowItem->mUIResourceName)>0)
	{
		_TRACE_UNLOAD_PLIST_(pWindowItem->mUIResourceName)	
	}
}

bool IWindow::ccTouchBegan(cocos2d::CCTouch* touche, cocos2d::CCEvent* event)
{
	return m_bModelWindow;
}

void IWindow::ccTouchMoved(cocos2d::CCTouch* touche, cocos2d::CCEvent* event)
{

}

void IWindow::ccTouchEnded(cocos2d::CCTouch* touche, cocos2d::CCEvent* event)
{

}

void IWindow::ccTouchCancel(cocos2d::CCTouch* touche, cocos2d::CCEvent* event)
{

}
